### PR TITLE
[Merged by Bors] - feat(NumberField/CanonicalEmbedding): define the `plusPart` of a set

### PR DESCRIPTION
**Pull request by xroblot**
Let `A` be a subset of the `mixedSpace K` of a number field `K`. We say that `A` is symmetric at real places if it satisfies:

`∀ x, x ∈ A ↔ (fun w ↦ |x.1 w|, x.2) ∈ A`

If `A` is measurable and symmetric at real places, we prove that
```
volume A = 2 ^ nrRealPlaces K * volume (plusPart A)
```
where `plusPart` is the subset of elements of `A` that are positive at all real places. 

This PR is part of the proof of the Analytic Class Number Formula.

---

- [x] depends on: #18234

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean [hunk 1]
```diff
@@ -237,6 +237,18 @@ instance : NoAtoms (volume : Measure (mixedSpace K)) := by
...
+open Classical in
+/-- The set of points in the mixedSpace that are equal to `0` at a fixed (real) place has
+volume zero. -/
+theorem volume_eq_zero (w : {w // IsReal w}) :
+    volume ({x : mixedSpace K | x.1 w = 0}) = 0 := by
+  let A : AffineSubspace ℝ (mixedSpace K) :=
+    Submodule.toAffineSubspace (Submodule.mk ⟨⟨{x | x.1 w = 0}, by aesop⟩, rfl⟩ (by aesop))
+  convert Measure.addHaar_affineSubspace volume A fun h ↦ ?_
+  have : 1 ∈ A := h ▸ Set.mem_univ _
+  simp [A] at this
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean:250]:
```suggestion
  simpa [A] using (h ▸ Set.mem_univ _ : 1 ∈ A)
```
If you want.

File: Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean [hunk 2]
```diff
@@ -862,6 +874,126 @@ theorem negAt_signSet_apply_of_isReal (x : mixedSpace K) (w : {w // IsReal w}) :
...
+ theorem pos_of_not_mem_negAt_plusPart {x : mixedSpace K} (hx : x ∈ negAt s '' (plusPart A))
+    {w : {w // IsReal w}} (hw : w ∉ s) :
+    0 < x.1 w := by
+  obtain ⟨y, hy, rfl⟩ := hx
+  rw [negAt_apply_of_isReal_and_not_mem _ hw]
+  exact hy.2 w
+
+ /-- The images of `plusPart` by `negAt` are pairwise disjoint. -/
+ theorem disjoint_negAt_plusPart : Pairwise (Disjoint on (fun s ↦ negAt s '' (plusPart A))) := by
+  classical
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean:906]:
```suggestion
```
This is not needed.

**Review by riccardobrasca — APPROVED** [review]
Thanks!

bors d+

File: Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean (hunk 1 above)
Comment (xroblot) [reply Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean:250]: Nice

File: Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean [hunk 3]
```diff
@@ -862,6 +874,126 @@ theorem negAt_signSet_apply_of_isReal (x : mixedSpace K) (w : {w // IsReal w}) :
...
+  convert_to MeasurableSet (A ∩ (⋂ w, {x | 0 < x.1 w}))
+  · ext; simp
+  · refine hm.inter (MeasurableSet.iInter fun _ ↦ ?_)
+    exact measurableSet_lt measurable_const ((measurable_pi_apply _).comp' measurable_fst)
+
+variable (s) in
+theorem measurableSet_negAt_plusPart (hm : MeasurableSet A) :
+    MeasurableSet (negAt s '' (plusPart A)) := by
+  rw [← negAt_preimage]
+  exact (measurableSet_plusPart hm).preimage (negAt s).continuous.measurable
```
[the remark this answers is not in the document]
Comment (riccardobrasca) [reply Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean:977]: This is a relatively new feature of `▸`: it works also in place of `rw [← ...]`

File: Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean [hunk 4]
```diff
@@ -862,6 +874,126 @@ theorem negAt_signSet_apply_of_isReal (x : mixedSpace K) (w : {w // IsReal w}) :
...
+/-- Assume that `A`  is symmetric at real places then, the union of the images of `plusPart`
+by `negAt` and of the set of elements of `A` that are zero at at least one real place
+is equal to `A`. -/
+theorem iUnion_negAt_plusPart_union :
+    (⋃ s, negAt s '' (plusPart A)) ∪ (A ∩ (⋃ w, {x | x.1 w = 0})) = A := by
+  ext x
+  rw [Set.mem_union, Set.mem_inter_iff, Set.mem_iUnion, Set.mem_iUnion]
+  refine ⟨?_, fun h ↦ ?_⟩
+  · rintro (⟨s, ⟨x, ⟨hx, _⟩, rfl⟩⟩ | h)
+    · simp_rw (config := {singlePass := true}) [hA, negAt_apply_abs_of_isReal, negAt_apply_snd]
```
[the remark this answers is not in the document]
Comment (xroblot) [reply Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean:942]: Without the `single pass`, it goes into an infinite recursion by rewriting using `hA` over and over

[the remark this answers is not in the document]
Comment (riccardobrasca) [reply Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean:938]: Ah, I guess I didn't notice the yellow bar didn't go away for me.

**Pull request closed without merging**